components/merkledb/src/proof_list_index/lazy.rs

```diff
@@ -0,0 +1,876 @@
+    /// index.push(1);
+    /// assert!(!index.is_empty());
+    ///
+    /// index.clear();
+    /// assert!(index.is_empty());
+    /// ```
+    pub fn clear(&mut self) {
+        self.base.clear();
+        self.state.clear();
+    }
+
+    pub fn update_hashes(&mut self) {
+        let hash = self.calculate_and_save_hashes();
+        let mut metadata = self.state.get();
+        metadata.update_hash(hash);
+        self.state.set(metadata);
+    }
+
+    fn calculate_and_save_hashes(&mut self) -> Hash {
+        let mut hashes: Vec<Hash> = self
+            .iter()
+            .map(|v| HashTag::hash_leaf(&v.to_bytes()))
+            .collect();
+
+        if hashes.is_empty() {
+            return HashTag::empty_list_hash();
+        }
+
+        for (index, hash) in hashes.iter().enumerate() {
+            let key = ProofListKey::new(1, index as u64);
+            self.set_branch(key, *hash);
+        }
+
+        let mut end = hashes.len();
+        let mut index = 0;
+        let mut height = 2;
+        let mut node_index = 0;
+
+        while end > 1 {
+            let first = hashes[index];
+
+            let result = if index < end - 1 {
+                HashTag::hash_node(&first, &hashes[index + 1])
+            } else {
+                HashTag::hash_single_node(&first)
+            };
+
+            hashes[index / 2] = result;
+
+            index += 2;
+
+            let key = ProofListKey::new(height, node_index as u64);
+            self.set_branch(key, result);
+
+            node_index += 1;
+            if index >= end {
+                index = 0;
+                node_index = 0;
+                end = end / 2 + end % 2;
+                height += 1;
+            }
+        }
+
+        HashTag::hash_list_node(hashes.len() as u64, hashes[0])
+    }
+}
+
+impl<T, V> ObjectHash for LazyListIndex<T, V>
+where
+    T: IndexAccess,
+    V: BinaryValue + ObjectHash + std::fmt::Debug,
+{
+    /// Returns a list hash of the proof list or a hash value of the empty list.
+    ///
+    /// List hash is calculated as follows:
+    /// ```text
+    /// h = sha-256( HashTag::List || len as u64 || merkle_root )
+    /// ```
+    /// Empty list hash:
+    /// ```text
+    /// h = sha-256( HashTag::List || 0 || Hash::default() )
+    /// ```
+    ///
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum_merkledb::{TemporaryDB, Database, ProofListIndex, HashTag, ObjectHash};
+    /// use exonum_crypto::Hash;
+    ///
+    /// let db = TemporaryDB::new();
+    /// let name = "name";
+    /// let fork = db.fork();
+    /// let mut index = ProofListIndex::new(name, &fork);
+    ///
+    /// let default_hash = index.object_hash();
+    /// assert_eq!(HashTag::empty_list_hash(), default_hash);
+    ///
+    /// index.push(1);
+    /// let hash = index.object_hash();
+    /// assert_ne!(hash, default_hash);
+    /// ```
+    fn object_hash(&self) -> Hash {
+        self.state.get().hash.expect("Hash is not calculated yet")
```

Comment:
That looks like a possibility to shoot yourself in the foot. Maybe if hash is required, it should be calculated right here right now? Or at least documentation should tell you that this method may panic.

Reply:
Good point, but to calculate hash we need to obtain a mutable reference to self. It's not possible in some cases, for example when we creating index from the snapshot.

Reply:
Hmm... You have a method `calculate_and_save_hashes`. You can add method `calculate_hashes` without saving, it would calculate hash once and won't store it anywhere. It wouldn't require mutable reference, would it?

However, there should be a note that if this method is calles without `update_hashes` it will be time-consuming. But at least it won't panic.

Reply:
You can rewrite index a little bit to make `calculate_and_save_hashes` immutable, thus for fork `object_hash` invocation will be update index but for snapshot you just only calculate object hash without saving.